pkg/core/blockchain.go

```diff
@@ -647,12 +656,37 @@ func (bc *Blockchain) Run() {
 		case <-bc.stopCh:
 			return
 		case <-persistTimer.C:
+			var oldPersisted uint32
+			var gcDur time.Duration
+
+			if bc.config.RemoveUntraceableBlocks {
+				oldPersisted = atomic.LoadUint32(&bc.persistedHeight) / bc.config.GarbageCollectionPeriod
+			}
 			dur, err := bc.persist(nextSync)
 			if err != nil {
 				bc.log.Warn("failed to persist blockchain", zap.Error(err))
 			}
+			if bc.config.RemoveUntraceableBlocks {
+				newPersisted := atomic.LoadUint32(&bc.persistedHeight) / bc.config.GarbageCollectionPeriod
+				var tgtBlock = int64(newPersisted)
+
+				tgtBlock -= int64(bc.config.MaxTraceableBlocks)
+				if bc.config.P2PStateExchangeExtensions {
+					syncP := newPersisted / uint32(bc.config.StateSyncInterval)
+					syncP--
+					syncP *= uint32(bc.config.StateSyncInterval)
+					if tgtBlock > int64(syncP) {
+						tgtBlock = int64(syncP)
+					}
```

Comment:
This differs from calculations for outdated blocks when P2PStateExchangeExtensions are on. There we have:
```
var start, stop uint32
if bc.config.P2PStateExchangeExtensions {
// remove batch of old blocks starting from P2-MaxTraceableBlocks-StateSyncInterval up to P2-MaxTraceableBlocks
if block.Index >= 2*uint32(bc.config.StateSyncInterval) &&
	block.Index >= uint32(bc.config.StateSyncInterval)+bc.config.MaxTraceableBlocks && // check this in case if MaxTraceableBlocks>StateSyncInterval
	int(block.Index)%bc.config.StateSyncInterval == 0 {
	stop = block.Index - uint32(bc.config.StateSyncInterval) - bc.config.MaxTraceableBlocks
	if stop > uint32(bc.config.StateSyncInterval) {
		start = stop - uint32(bc.config.StateSyncInterval)
	}
}
}
```
So given the latest state sync points P1 and P2 and the newest one P0 we remove old blocks up to P2-MaxTraceableBlocks.

And IIUC, here you don't take into account MaxTraceableBlocks and run GC to clean MPT data up to P2 height (if it's lower than tgtBlock).

Reply:
You're right, we don't need them.